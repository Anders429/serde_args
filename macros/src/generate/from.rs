use crate::Container;
use proc_macro2::{
    Delimiter,
    Group,
    Span,
    TokenStream,
};
use quote::quote;
use syn::{
    punctuated::Punctuated,
    token::{
        Brace,
        Paren,
    },
    FieldPat,
    Fields,
    Ident,
    ItemEnum,
    ItemStruct,
    Lit,
    LitInt,
    Member,
    Pat,
    PatIdent,
    PatLit,
    PatStruct,
    PatTupleStruct,
    Type,
};

fn collect_variant_patterns(item: &ItemEnum) -> impl Iterator<Item = Pat> + '_ {
    item.variants.iter().map(|variant| match &variant.fields {
        fields @ Fields::Named(_) => {
            let fields = fields.iter().enumerate().map(|(index, field)| {
                let member = match &field.ident {
                    Some(ident) => Member::Named(ident.clone()),
                    None => Member::Unnamed(index.into()),
                };
                let pat = match &member {
                    Member::Named(ident) => Pat::Ident(PatIdent {
                        attrs: vec![],
                        by_ref: None,
                        mutability: None,
                        ident: ident.clone(),
                        subpat: None,
                    }),
                    Member::Unnamed(index) => Pat::Lit(PatLit {
                        attrs: vec![],
                        lit: Lit::Int(LitInt::new(&format!("{}", index.index), Span::call_site())),
                    }),
                };
                FieldPat {
                    attrs: vec![],
                    member,
                    colon_token: None,
                    pat: Box::new(pat),
                }
            });
            let punctuated_fields = fields.clone().collect::<Punctuated<_, _>>();
            let group = Group::new(Delimiter::Brace, quote!(#(#fields),*));
            Pat::Struct(PatStruct {
                attrs: vec![],
                qself: None,
                path: variant.ident.clone().into(),
                brace_token: Brace {
                    span: group.delim_span(),
                },
                fields: punctuated_fields,
                rest: None,
            })
        }
        fields @ Fields::Unnamed(_) => {
            let elems = (0..fields.len()).map(|index| {
                Pat::Ident(PatIdent {
                    attrs: vec![],
                    by_ref: None,
                    mutability: None,
                    ident: Ident::new(&format!("__{}", index), Span::call_site()),
                    subpat: None,
                })
            });
            let punctuated_elems = elems.clone().collect::<Punctuated<_, _>>();
            let group = Group::new(Delimiter::Parenthesis, quote!(#(#elems),*));
            Pat::TupleStruct(PatTupleStruct {
                attrs: vec![],
                qself: None,
                path: variant.ident.clone().into(),
                paren_token: Paren {
                    span: group.delim_span(),
                },
                elems: punctuated_elems,
            })
        }
        Fields::Unit => Pat::Ident(PatIdent {
            attrs: vec![],
            by_ref: None,
            mutability: None,
            ident: variant.ident.clone(),
            subpat: None,
        }),
    })
}

fn collect_field_members(item: &ItemStruct) -> impl Iterator<Item = Member> + '_ {
    item.fields
        .iter()
        .enumerate()
        .map(|(index, field)| match &field.ident {
            Some(ident) => Member::Named(ident.clone()),
            None => Member::Unnamed(index.into()),
        })
}

pub(crate) fn from_newtype_to_container(
    container: &Container,
    from: &Type,
    to: &Type,
) -> TokenStream {
    let ident = container.identifier();
    match container {
        Container::Enum(item) => {
            // Prepare the variants.
            let variants = collect_variant_patterns(item)
                .map(|pattern| quote! {#ident::#pattern => #to::#pattern,});
            quote! {
                impl ::std::convert::From<#from> for #to {
                    fn from(from: #from) -> #to {
                        match from.0 {
                            #(#variants)*
                        }
                    }
                }
            }
        }
        Container::Struct(item) => {
            // Prepare the fields.
            let fields = collect_field_members(&item).map(|ident| quote!(#ident: from.0.#ident));
            quote! {
                impl ::std::convert::From<#from> for #to {
                    fn from(from: #from) -> #to {
                        #to {
                            #(#fields),*
                        }
                    }
                }
            }
        }
    }
}

pub(crate) fn from_container_to_newtype(
    container: &Container,
    from: &Type,
    to: &Type,
) -> TokenStream {
    let ident = container.identifier();
    match container {
        Container::Enum(item) => {
            // Prepare the variants.
            let variants = collect_variant_patterns(item)
                .map(|pattern| quote!(#from::#pattern => #to(#ident::#pattern),));
            quote! {
                impl ::std::convert::From<#from> for #to {
                    fn from(from: #from) -> #to {
                        match from {
                            #(#variants)*
                        }
                    }
                }
            }
        }
        Container::Struct(item) => {
            // Prepare the fields.
            let fields = collect_field_members(&item).map(|ident| quote!(#ident: from.#ident));
            quote! {
                impl ::std::convert::From<#from> for #to {
                    fn from(from: #from) -> #to {
                        #to(#ident {
                            #(#fields),*
                        })
                    }
                }
            }
        }
    }
}

pub(crate) fn from_foreign_to_container(
    container: &Container,
    from: &Type,
    intermediate_a: &Type,
    intermediate_b: &Type,
    to: &Type,
) -> TokenStream {
    match container {
        Container::Enum(item) => {
            // Prepare the variants.
            let variants = collect_variant_patterns(item)
                .map(|pattern| quote!(#intermediate_b::#pattern => #to::#pattern,));
            quote! {
                impl ::std::convert::From<#from> for #to {
                    fn from(from: #from) -> #to {
                        let converted_from = #intermediate_b::from(#intermediate_a::from(from));
                        match converted_from.0 {
                            #(#variants)*
                        }
                    }
                }
            }
        }
        Container::Struct(item) => {
            // Prepare the fields.
            let fields =
                collect_field_members(&item).map(|ident| quote!(#ident: converted_from.0.#ident));
            quote! {
                impl ::std::convert::From<#from> for #to {
                    fn from(from: #from) -> #to {
                        let converted_from = #intermediate_b::from(#intermediate_a::from(from));
                        #to {
                            #(#fields),*
                        }
                    }
                }
            }
        }
    }
}

pub(crate) fn from_container_to_foreign(
    from: &Type,
    intermediate_a: &Type,
    intermediate_b: &Type,
    to: &Type,
) -> TokenStream {
    quote! {
        impl ::std::convert::From<#from> for #to {
            fn from(from: #from) -> #to {
                #to::from(#intermediate_b::from(#intermediate_a::from(from)))
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::{
        collect_field_members,
        collect_variant_patterns,
        from_container_to_newtype,
        from_foreign_to_container,
        from_newtype_to_container,
    };
    use crate::test::SinglePattern;
    use claims::{
        assert_ok,
        assert_ok_eq,
    };
    use syn::{
        parse2 as parse,
        parse_str,
        File,
    };

    #[test]
    fn collect_variant_patterns_empty() {
        let item = assert_ok!(parse_str("enum Foo {}"));
        assert_eq!(collect_variant_patterns(&item).collect::<Vec<_>>(), vec![]);
    }

    #[test]
    fn collect_variant_patterns_unit() {
        let item = assert_ok!(parse_str(
            "enum Foo {
                Bar,
            }"
        ));
        assert_eq!(
            collect_variant_patterns(&item).collect::<Vec<_>>(),
            vec![assert_ok!(parse_str::<SinglePattern>("Bar")).0]
        );
    }

    #[test]
    fn collect_variant_patterns_newtype() {
        let item = assert_ok!(parse_str(
            "enum Foo {
                Bar(usize),
            }"
        ));
        assert_eq!(
            collect_variant_patterns(&item).collect::<Vec<_>>(),
            vec![assert_ok!(parse_str::<SinglePattern>("Bar(__0)")).0]
        );
    }

    #[test]
    fn collect_variant_patterns_tuple() {
        let item = assert_ok!(parse_str(
            "enum Foo {
                Bar(usize, u8, String, bool),
            }"
        ));
        assert_eq!(
            collect_variant_patterns(&item).collect::<Vec<_>>(),
            vec![assert_ok!(parse_str::<SinglePattern>("Bar(__0, __1, __2, __3)")).0]
        );
    }

    #[test]
    fn collect_variant_patterns_struct() {
        let item = assert_ok!(parse_str(
            "enum Foo {
                Bar {
                    baz: usize,
                    qux: bool,
                    quux: String,
                },
            }"
        ));
        assert_eq!(
            collect_variant_patterns(&item).collect::<Vec<_>>(),
            vec![assert_ok!(parse_str::<SinglePattern>("Bar {baz, qux, quux}")).0]
        );
    }

    #[test]
    fn collect_field_members_empty() {
        let item = assert_ok!(parse_str("struct Foo {}"));
        assert_eq!(collect_field_members(&item).collect::<Vec<_>>(), vec![]);
    }

    #[test]
    fn collect_field_members_single() {
        let item = assert_ok!(parse_str(
            "struct Foo {
                bar: usize,
            }"
        ));
        assert_eq!(
            collect_field_members(&item).collect::<Vec<_>>(),
            vec![assert_ok!(parse_str("bar"))]
        );
    }

    #[test]
    fn collect_field_members_multiple() {
        let item = assert_ok!(parse_str(
            "struct Foo {
                bar: usize,
                baz: String,
                qux: u8,
            }"
        ));
        assert_eq!(
            collect_field_members(&item).collect::<Vec<_>>(),
            vec![
                assert_ok!(parse_str("bar")),
                assert_ok!(parse_str("baz")),
                assert_ok!(parse_str("qux"))
            ]
        );
    }

    #[test]
    fn from_newtype_to_container_enum_no_variants() {
        assert_ok_eq!(
            parse::<File>(from_newtype_to_container(
                &assert_ok!(parse_str("enum Foo {}")),
                &assert_ok!(parse_str("Bar<Foo>")),
                &assert_ok!(parse_str("Baz"))
            )),
            assert_ok!(parse_str(
                "
            impl ::std::convert::From<Bar<Foo>> for Baz {
                fn from(from: Bar<Foo>) -> Baz {
                    match from.0 {}
                }
            }"
            ))
        );
    }

    #[test]
    fn from_newtype_to_container_enum() {
        assert_ok_eq!(
            parse::<File>(from_newtype_to_container(
                &assert_ok!(parse_str(
                    "enum Foo {
                        Bar,
                        Baz(usize),
                        Qux {
                            quux: String,
                        }
                    }"
                )),
                &assert_ok!(parse_str("A<Foo>")),
                &assert_ok!(parse_str("B"))
            )),
            assert_ok!(parse_str(
                "
            impl ::std::convert::From<A<Foo>> for B {
                fn from(from: A<Foo>) -> B {
                    match from.0 {
                        Foo::Bar => B::Bar,
                        Foo::Baz(__0) => B::Baz(__0),
                        Foo::Qux {quux} => B::Qux {quux},
                    }
                }
            }"
            ))
        );
    }

    #[test]
    fn from_newtype_to_container_enum_with_attributes() {
        assert_ok_eq!(
            parse::<File>(from_newtype_to_container(
                &assert_ok!(parse_str(
                    "enum Foo {
                        /// Bar documentation.
                        Bar,
                        /// Baz documentation.
                        Baz(usize),
                        /// Qux documentation.
                        Qux {
                            /// Quux documentation.
                            quux: String,
                        }
                    }"
                )),
                &assert_ok!(parse_str("A<Foo>")),
                &assert_ok!(parse_str("B"))
            )),
            assert_ok!(parse_str(
                "
            impl ::std::convert::From<A<Foo>> for B {
                fn from(from: A<Foo>) -> B {
                    match from.0 {
                        Foo::Bar => B::Bar,
                        Foo::Baz(__0) => B::Baz(__0),
                        Foo::Qux {quux} => B::Qux {quux},
                    }
                }
            }"
            ))
        );
    }

    #[test]
    fn from_newtype_to_container_struct_no_fields() {
        assert_ok_eq!(
            parse::<File>(from_newtype_to_container(
                &assert_ok!(parse_str("struct Foo {}")),
                &assert_ok!(parse_str("Bar<Foo>")),
                &assert_ok!(parse_str("Baz"))
            )),
            assert_ok!(parse_str(
                "
            impl ::std::convert::From<Bar<Foo>> for Baz {
                fn from(from: Bar<Foo>) -> Baz {
                    Baz {}
                }
            }"
            ))
        );
    }

    #[test]
    fn from_newtype_to_container_struct() {
        assert_ok_eq!(
            parse::<File>(from_newtype_to_container(
                &assert_ok!(parse_str(
                    "struct Foo {
                        bar: usize,
                        baz: String,
                    }"
                )),
                &assert_ok!(parse_str("Bar<Foo>")),
                &assert_ok!(parse_str("Baz"))
            )),
            assert_ok!(parse_str(
                "
            impl ::std::convert::From<Bar<Foo>> for Baz {
                fn from(from: Bar<Foo>) -> Baz {
                    Baz {
                        bar: from.0.bar,
                        baz: from.0.baz
                    }
                }
            }"
            ))
        );
    }

    #[test]
    fn from_newtype_to_container_struct_with_attributes() {
        assert_ok_eq!(
            parse::<File>(from_newtype_to_container(
                &assert_ok!(parse_str(
                    "
                    /// Foo documentation.
                    struct Foo {
                        /// Bar documentation.
                        bar: usize,
                        /// Baz documentation.
                        baz: String,
                    }"
                )),
                &assert_ok!(parse_str("Bar<Foo>")),
                &assert_ok!(parse_str("Baz"))
            )),
            assert_ok!(parse_str(
                "
            impl ::std::convert::From<Bar<Foo>> for Baz {
                fn from(from: Bar<Foo>) -> Baz {
                    Baz {
                        bar: from.0.bar,
                        baz: from.0.baz
                    }
                }
            }"
            ))
        );
    }

    #[test]
    fn from_container_to_newtype_enum_no_variants() {
        assert_ok_eq!(
            parse::<File>(from_container_to_newtype(
                &assert_ok!(parse_str("enum Foo {}")),
                &assert_ok!(parse_str("Baz")),
                &assert_ok!(parse_str("Bar<Foo>")),
            )),
            assert_ok!(parse_str(
                "
            impl ::std::convert::From<Baz> for Bar<Foo> {
                fn from(from: Baz) -> Bar<Foo> {
                    match from {}
                }
            }"
            ))
        );
    }

    #[test]
    fn from_container_to_newtype_enum() {
        assert_ok_eq!(
            parse::<File>(from_container_to_newtype(
                &assert_ok!(parse_str(
                    "enum Foo {
                        Bar,
                        Baz(usize),
                        Qux {
                            quux: String,
                        }
                    }"
                )),
                &assert_ok!(parse_str("A")),
                &assert_ok!(parse_str("B::<Foo>"))
            )),
            assert_ok!(parse_str(
                "
            impl ::std::convert::From<A> for B::<Foo> {
                fn from(from: A) -> B::<Foo> {
                    match from {
                        A::Bar => B::<Foo>(Foo::Bar),
                        A::Baz(__0) => B::<Foo>(Foo::Baz(__0)),
                        A::Qux {quux} => B::<Foo>(Foo::Qux {quux}),
                    }
                }
            }"
            ))
        );
    }

    #[test]
    fn from_container_to_newtype_enum_with_attributes() {
        assert_ok_eq!(
            parse::<File>(from_container_to_newtype(
                &assert_ok!(parse_str(
                    "enum Foo {
                        /// Bar documentation.
                        Bar,
                        /// Baz documentation.
                        Baz(usize),
                        /// Qux documentation.
                        Qux {
                            /// Quux documentation.
                            quux: String,
                        }
                    }"
                )),
                &assert_ok!(parse_str("A")),
                &assert_ok!(parse_str("B::<Foo>"))
            )),
            assert_ok!(parse_str(
                "
            impl ::std::convert::From<A> for B::<Foo> {
                fn from(from: A) -> B::<Foo> {
                    match from {
                        A::Bar => B::<Foo>(Foo::Bar),
                        A::Baz(__0) => B::<Foo>(Foo::Baz(__0)),
                        A::Qux {quux} => B::<Foo>(Foo::Qux {quux}),
                    }
                }
            }"
            ))
        );
    }

    #[test]
    fn from_container_to_newtype_struct_no_fields() {
        assert_ok_eq!(
            parse::<File>(from_container_to_newtype(
                &assert_ok!(parse_str("struct Foo {}")),
                &assert_ok!(parse_str("Bar")),
                &assert_ok!(parse_str("Baz::<Foo>"))
            )),
            assert_ok!(parse_str(
                "
            impl ::std::convert::From<Bar> for Baz::<Foo> {
                fn from(from: Bar) -> Baz::<Foo> {
                    Baz::<Foo>(Foo {})
                }
            }"
            ))
        );
    }

    #[test]
    fn from_container_to_newtype_struct() {
        assert_ok_eq!(
            parse::<File>(from_container_to_newtype(
                &assert_ok!(parse_str(
                    "struct Foo {
                        bar: usize,
                        baz: String,
                    }"
                )),
                &assert_ok!(parse_str("Bar")),
                &assert_ok!(parse_str("Baz::<Foo>"))
            )),
            assert_ok!(parse_str(
                "
            impl ::std::convert::From<Bar> for Baz::<Foo> {
                fn from(from: Bar) -> Baz::<Foo> {
                    Baz::<Foo>(Foo {
                        bar: from.bar,
                        baz: from.baz
                    })
                }
            }"
            ))
        );
    }

    #[test]
    fn from_container_to_newtype_struct_with_attributes() {
        assert_ok_eq!(
            parse::<File>(from_container_to_newtype(
                &assert_ok!(parse_str(
                    "
                    /// Foo documentation.
                    struct Foo {
                        /// Bar documentation.
                        bar: usize,
                        /// Baz documentation.
                        baz: String,
                    }"
                )),
                &assert_ok!(parse_str("Bar")),
                &assert_ok!(parse_str("Baz::<Foo>"))
            )),
            assert_ok!(parse_str(
                "
            impl ::std::convert::From<Bar> for Baz::<Foo> {
                fn from(from: Bar) -> Baz::<Foo> {
                    Baz::<Foo>(Foo {
                        bar: from.bar,
                        baz: from.baz
                    })
                }
            }"
            ))
        );
    }

    #[test]
    fn from_foreign_to_container_struct_empty() {
        assert_ok_eq!(
            parse::<File>(from_foreign_to_container(
                &assert_ok!(parse_str("struct Foo {}")),
                &assert_ok!(parse_str("Bar")),
                &assert_ok!(parse_str("Foo")),
                &assert_ok!(parse_str("Baz")),
                &assert_ok!(parse_str("Qux")),
            )),
            assert_ok!(parse_str(
                "
            impl ::std::convert::From<Bar> for Qux {
                fn from(from: Bar) -> Qux {
                    let converted_from = Baz::from(Foo::from(from));
                    Qux {}
                }
            }"
            ))
        );
    }

    #[test]
    fn from_foreign_to_container_struct() {
        assert_ok_eq!(
            parse::<File>(from_foreign_to_container(
                &assert_ok!(parse_str(
                    "struct Foo {
                        bar: usize,
                        baz: String,
                    }"
                )),
                &assert_ok!(parse_str("Bar")),
                &assert_ok!(parse_str("Baz")),
                &assert_ok!(parse_str("Foo")),
                &assert_ok!(parse_str("Qux")),
            )),
            assert_ok!(parse_str(
                "
            impl ::std::convert::From<Bar> for Qux {
                fn from(from: Bar) -> Qux {
                    let converted_from = Foo::from(Baz::from(from));
                    Qux {
                        bar: converted_from.0.bar,
                        baz: converted_from.0.baz
                    }
                }
            }"
            ))
        );
    }

    #[test]
    fn from_foreign_to_container_enum_empty() {
        assert_ok_eq!(
            parse::<File>(from_foreign_to_container(
                &assert_ok!(parse_str("enum Foo {}")),
                &assert_ok!(parse_str("Bar")),
                &assert_ok!(parse_str("Foo")),
                &assert_ok!(parse_str("Baz")),
                &assert_ok!(parse_str("Qux")),
            )),
            assert_ok!(parse_str(
                "
            impl ::std::convert::From<Bar> for Qux {
                fn from(from: Bar) -> Qux {
                    let converted_from = Baz::from(Foo::from(from));
                    match converted_from.0 {}
                }
            }"
            ))
        );
    }

    #[test]
    fn from_foreign_to_container_enum() {
        assert_ok_eq!(
            parse::<File>(from_foreign_to_container(
                &assert_ok!(parse_str(
                    "enum Foo {
                        Unit,
                        Newtype(usize),
                        Tuple(usize, u8, String),
                        Struct {
                            a: usize,
                            b: bool,
                        }
                    }"
                )),
                &assert_ok!(parse_str("Bar")),
                &assert_ok!(parse_str("Foo")),
                &assert_ok!(parse_str("Baz")),
                &assert_ok!(parse_str("Qux")),
            )),
            assert_ok!(parse_str(
                "
            impl ::std::convert::From<Bar> for Qux {
                fn from(from: Bar) -> Qux {
                    let converted_from = Baz::from(Foo::from(from));
                    match converted_from.0 {
                        Baz::Unit => Qux::Unit,
                        Baz::Newtype(__0) => Qux::Newtype(__0),
                        Baz::Tuple(__0, __1, __2) => Qux::Tuple(__0, __1, __2),
                        Baz::Struct {a, b} => Qux::Struct {a, b},
                    }
                }
            }"
            ))
        );
    }

    #[test]
    fn from_container_to_foreign() {
        assert_ok_eq!(
            parse::<File>(super::from_container_to_foreign(
                &assert_ok!(parse_str("Foo")),
                &assert_ok!(parse_str("Bar")),
                &assert_ok!(parse_str("Baz")),
                &assert_ok!(parse_str("Qux")),
            )),
            assert_ok!(parse_str(
                "
            impl ::std::convert::From<Foo> for Qux {
                fn from(from: Foo) -> Qux {
                    Qux::from(Baz::from(Bar::from(from)))
                }
            }
            "
            ))
        );
    }
}
