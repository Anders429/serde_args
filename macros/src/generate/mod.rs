//! Generating the actual code.

mod from;

pub(crate) use from::from;

use crate::{container::Descriptions, Container};
use proc_macro2::{Delimiter, Group, Literal, Punct, Spacing, Span, TokenStream, TokenTree};
use quote::{quote, ToTokens};
use std::iter;
use syn::{
    token::Bracket, token::Paren, AttrStyle, Attribute, Ident, MacroDelimiter, Meta, MetaList,
    Path, PathArguments, PathSegment, Token, Visibility,
};

fn push_serde_attribute(attrs: &mut Vec<Attribute>, meta_tokens: TokenStream) {
    let meta_group = Group::new(Delimiter::Parenthesis, meta_tokens);
    let meta = Meta::List(MetaList {
        path: Path {
            leading_colon: None,
            segments: iter::once(PathSegment {
                ident: Ident::new("serde", Span::call_site()),
                arguments: PathArguments::None,
            })
            .collect(),
        },
        delimiter: MacroDelimiter::Paren(Paren {
            span: meta_group.delim_span(),
        }),
        tokens: meta_group.stream(),
    });

    let mut tokens = TokenStream::new();
    meta.to_tokens(&mut tokens);
    let group = Group::new(Delimiter::Bracket, tokens);

    attrs.push(Attribute {
        pound_token: Token![#](Span::call_site()),
        style: AttrStyle::Outer,
        bracket_token: Bracket {
            span: group.delim_span(),
        },
        meta,
    });
}

pub(crate) fn phase_1(mut container: Container, ident: &Ident) -> Container {
    let attribute_tokens: TokenStream = [
        TokenTree::Ident(Ident::new("rename", Span::call_site())),
        TokenTree::Punct(Punct::new('=', Spacing::Alone)),
        TokenTree::Literal(Literal::string(&format!("{}", ident))),
    ]
    .into_iter()
    .collect();
    match &mut container {
        Container::Enum(item) => {
            push_serde_attribute(&mut item.attrs, attribute_tokens);
            item.vis = Visibility::Inherited;
            item.ident = Ident::new("Phase1", Span::call_site());
        }
        Container::Struct(item) => {
            push_serde_attribute(&mut item.attrs, attribute_tokens);
            item.vis = Visibility::Inherited;
            item.ident = Ident::new("Phase1", Span::call_site());
        }
    };

    container
}

pub(crate) fn phase_2(
    mut container: Container,
    descriptions: Descriptions,
    ident: &Ident,
) -> TokenStream {
    // Remove all attributes from this item.
    match &mut container {
        Container::Enum(item) => {
            item.attrs.clear();
            item.vis = Visibility::Inherited;
            item.ident = Ident::new("Phase2", Span::call_site());
            item.variants.iter_mut().for_each(|variant| {
                variant.attrs.clear();
                variant
                    .fields
                    .iter_mut()
                    .for_each(|field| field.attrs.clear());
            });
        }
        Container::Struct(item) => {
            item.attrs.clear();
            item.vis = Visibility::Inherited;
            item.ident = Ident::new("Phase2", Span::call_site());
            item.fields.iter_mut().for_each(|field| field.attrs.clear());
        }
    };

    // Define a `From` implementation from Phase 1.
    let from = from(
        &container,
        &Ident::new("Phase1", Span::call_site()),
        &Ident::new("Phase2", Span::call_site()),
    );

    // Define the `expecting()` match statements.
    let container_exprs = descriptions
        .container
        .exprs
        .into_iter()
        .map(|expr| quote!(formatter.write_str(#expr)?;));
    let key_exprs = descriptions
        .keys
        .into_iter()
        .enumerate()
        .map(|(index, documentation)| {
            let documentation_exprs = documentation
                .exprs
                .into_iter()
                .map(|expr| quote!(formatter.write_str(#expr)?;));
            quote!(Some(#index) => {#(#documentation_exprs)*})
        });

    let ident_string = format!("{}", ident);
    quote! {
        #container
        #from

        impl<'de> ::serde::de::Deserialize<'de> for Phase2 {
            fn deserialize<D>(deserializer: D) -> Result<Phase2, D::Error> where D: ::serde::de::Deserializer<'de> {
                struct Phase2Visitor;

                impl<'de> ::serde::de::Visitor<'de> for Phase2Visitor {
                    type Value = Phase2;

                    fn expecting(&self, formatter: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
                        match formatter.width() {
                            #(#key_exprs)*
                            _ => {#(#container_exprs)*}
                        }
                        Ok(())
                    }

                    fn visit_newtype_struct<D>(self, deserializer: D) -> Result<Self::Value, D::Error> where D: ::serde::de::Deserializer<'de> {
                        use ::serde::de::Deserialize;
                        Phase1::deserialize(deserializer).map(Into::into)
                    }
                }

                deserializer.deserialize_newtype_struct(#ident_string, Phase2Visitor)
            }
        }
    }
}

pub(crate) fn phase_3(mut container: Container) -> TokenStream {
    // Insert the `serde(from)` attribute.
    let from_tokens: TokenStream = [
        TokenTree::Ident(Ident::new("from", Span::call_site())),
        TokenTree::Punct(Punct::new('=', Spacing::Alone)),
        TokenTree::Literal(Literal::string("Phase2")),
    ]
    .into_iter()
    .collect();
    match &mut container {
        Container::Enum(item) => {
            push_serde_attribute(&mut item.attrs, from_tokens);
            item.vis = Visibility::Public(Token!(pub)(Span::call_site()));
        }
        Container::Struct(item) => {
            push_serde_attribute(&mut item.attrs, from_tokens);
            item.vis = Visibility::Public(Token!(pub)(Span::call_site()));
        }
    };
    let ident = container.identifier();

    // Create a `From` implementation.
    let from = from(&container, &Ident::new("Phase2", Span::call_site()), ident);

    quote! {
        #container
        #from
    }
}

#[cfg(test)]
mod tests {
    use super::push_serde_attribute;
    use crate::test::OuterAttributes;
    use claims::assert_ok;
    use proc_macro2::{Ident, Span, TokenTree};
    use std::iter;
    use syn::parse_str;

    #[test]
    fn push_serde_attribute_empty() {
        let mut attributes = vec![];

        push_serde_attribute(
            &mut attributes,
            iter::once(TokenTree::Ident(Ident::new("foo", Span::call_site()))).collect(),
        );

        assert_eq!(
            attributes,
            assert_ok!(parse_str::<OuterAttributes>("#[serde(foo)]")).0
        );
    }

    #[test]
    fn push_serde_attribute_nonempty() {
        let mut attributes = assert_ok!(parse_str::<OuterAttributes>("#[foo] #[bar]")).0;

        push_serde_attribute(
            &mut attributes,
            iter::once(TokenTree::Ident(Ident::new("foo", Span::call_site()))).collect(),
        );

        assert_eq!(
            attributes,
            assert_ok!(parse_str::<OuterAttributes>("#[foo] #[bar] #[serde(foo)]")).0
        );
    }
}
