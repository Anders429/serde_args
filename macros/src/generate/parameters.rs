use proc_macro2::Span;
use syn::{
    parse,
    parse::{
        Parse,
        ParseStream,
    },
    punctuated::Punctuated,
    Ident,
    Path,
    Token,
};

#[derive(Debug, Eq, Hash, PartialEq)]
pub(super) enum Parameter {
    DocHelp,
    Version,
}

#[derive(Debug, Eq, PartialEq)]
pub(super) struct Parameters(u8);

impl Parameters {
    #[cfg(test)]
    const EMPTY: u8 = 0;
    const VERSION: u8 = 1;
    // DocHelp must be the last one returned in iteration.
    const DOC_HELP: u8 = 2;
}

impl Parse for Parameters {
    fn parse(input: ParseStream) -> parse::Result<Self> {
        let mut parameters = 0;
        for path in Punctuated::<Path, Token![,]>::parse_terminated(input)? {
            let ident = path.require_ident()?;
            if *ident == Ident::new("doc_help", Span::call_site()) {
                parameters |= Parameters::DOC_HELP;
            } else if *ident == Ident::new("version", Span::call_site()) {
                parameters |= Parameters::VERSION;
            } else {
                return Err(syn::Error::new_spanned(
                    ident,
                    "invalid parameter; expected one of `doc_help` or `version`",
                ));
            }
        }
        Ok(Self(parameters))
    }
}

impl IntoIterator for Parameters {
    type Item = Parameter;
    type IntoIter = Iter;

    fn into_iter(self) -> Self::IntoIter {
        Iter { parameters: self.0 }
    }
}

pub(super) struct Iter {
    parameters: u8,
}

impl Iterator for Iter {
    type Item = Parameter;

    fn next(&mut self) -> Option<Self::Item> {
        if self.parameters & Parameters::VERSION != 0 {
            self.parameters ^= Parameters::VERSION;
            Some(Parameter::Version)
        } else if self.parameters & Parameters::DOC_HELP != 0 {
            self.parameters ^= Parameters::DOC_HELP;
            Some(Parameter::DocHelp)
        } else {
            None
        }
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        let size = self.parameters.count_ones() as usize;
        (size, Some(size))
    }
}

impl ExactSizeIterator for Iter {}

#[cfg(test)]
mod tests {
    use super::{
        Parameter,
        Parameters,
    };
    use claims::{
        assert_err,
        assert_ok_eq,
    };
    use syn::parse_str;

    #[test]
    fn parse_empty() {
        assert_ok_eq!(parse_str::<Parameters>(""), Parameters(Parameters::EMPTY));
    }

    #[test]
    fn parse_doc_help() {
        assert_ok_eq!(
            parse_str::<Parameters>("doc_help"),
            Parameters(Parameters::DOC_HELP)
        );
    }

    #[test]
    fn parse_version() {
        assert_ok_eq!(
            parse_str::<Parameters>("version"),
            Parameters(Parameters::VERSION)
        );
    }

    #[test]
    fn parse_all() {
        assert_ok_eq!(
            parse_str::<Parameters>("doc_help, version"),
            Parameters(Parameters::DOC_HELP | Parameters::VERSION)
        );
    }

    #[test]
    fn parse_unknown() {
        assert_eq!(
            format!("{}", assert_err!(parse_str::<Parameters>("unknown"))),
            "invalid parameter; expected one of `doc_help` or `version`"
        );
    }

    #[test]
    fn iter_none() {
        assert_eq!(
            Parameters(Parameters::EMPTY)
                .into_iter()
                .collect::<Vec<_>>(),
            &[]
        );
    }

    #[test]
    fn iter_doc_help() {
        assert_eq!(
            Parameters(Parameters::DOC_HELP)
                .into_iter()
                .collect::<Vec<_>>(),
            &[Parameter::DocHelp]
        );
    }

    #[test]
    fn iter_version() {
        assert_eq!(
            Parameters(Parameters::VERSION)
                .into_iter()
                .collect::<Vec<_>>(),
            &[Parameter::Version]
        );
    }

    #[test]
    fn iter_version_doc_help() {
        // `DocHelp` should always come last.
        // This is because the `DocHelp` `expecting()` function will never return `false`.
        assert_eq!(
            Parameters(Parameters::DOC_HELP | Parameters::VERSION)
                .into_iter()
                .collect::<Vec<_>>(),
            &[Parameter::Version, Parameter::DocHelp]
        );
    }
}
