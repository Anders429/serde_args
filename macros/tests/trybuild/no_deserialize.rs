use serde_args_macros::generate;

#[generate(doc_help)]
struct Foo {
    bar: usize,
    baz: String,
}

fn main() {}
