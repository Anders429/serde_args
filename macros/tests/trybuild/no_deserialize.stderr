error: cannot find attribute `serde` in this scope
 --> tests/trybuild/no_deserialize.rs:3:1
  |
3 | #[generate(doc_help)]
  | ^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `serde` is in scope, but it is a crate, not an attribute
  = note: this error originates in the attribute macro `generate` (in Nightly builds, run with -Z macro-backtrace for more info)
